plugins/newsdownloader.koplugin/epubdownloadbackend.lua

```diff
@@ -24,6 +24,48 @@ local max_redirects = 5; --prevent infinite redirects
 local TIMEOUT_CODE = "timeout" -- from socket.lua
 local MAXTIME_CODE = "maxtime reached" -- from sink_table_with_maxtime
 
+-- filter HTML using CSS selector
+local function filter(text, element)
+    local htmlparser = require("htmlparser")
+    local root = htmlparser.parse(text)
+    local filtered = ""
+    local selectors = {
+        "main",
+        "article",
+        "div#main",
+        "#main-article",
+        ".main-content",
+        "#body",
+        "#content",
+        ".content",
+        "div#article",
+        "div.article",
+        "div.post",
+        "div.post-outer",
+        ".l-root",
+        ".content-container",
+        ".StandardArticleBody_body",
+        "div#article-inner",
+        "div#newsstorytext",
+        "div.general",
+        }
+    if element then selectors[0] = element end
+    for _, sel in pairs(selectors) do
+       local elements = root:select(sel)
+       local stop = false
+       for _, e in ipairs(elements) do
+           filtered = e:getcontent()
+           if filtered then 
+               stop = true
+               break
+           end
+       end
+       if stop then break end
+    end
+    if not filtered then return text end
```

Comment:
Your initial assignment of the empty string `local filtered = ""` will make that test fail (so, you won't return text if there's no elements).
`if not filtered then ` will only trigger if filtered is `false` or `nil`. And not when it is an empty string `""` or an empty table `{}`. (That's different from most other languages.)

Reply:
Huh, that's again quite a large difference.